### PR TITLE
Load assembly from module base path before trying to load from GAC

### PR DESCRIPTION
**Pull request by adityapatwardhan**
## PR Summary

When a binary module has the module assembly in GAC, we load the assembly from GAC before trying to load it from module base path. 

This change attempts to load it from module base path before looking up in GAC.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [X] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by adityapatwardhan** [comment]
@SteveL-MSFT  I guess this would technically be a breaking change if someone had a assembly with the same name in the module base folder and are relying on loading it from GAC instead. But, I believe that should be a very rare occurance. Your thoughts?

**Comment by SteveL-MSFT** [comment]
@adityapatwardhan I can't imagine a scenario where a module author has an assembly in their module base folder but expects the assembly from the GAC to be loaded

**Review by TravisEz13 — APPROVED** [review]

**Comment by TravisEz13** [comment]
Technically it's breaking, let's document it as breaking, but I don't think this needs committee review

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 1]
```diff
@@ -4582,7 +4582,7 @@ internal string FixupFileName(string moduleBase, string name, string extension)
             if (!string.IsNullOrEmpty(ext) && ext.Equals(".dll", StringComparison.OrdinalIgnoreCase))
             {
                 Exception ignored = null;
-                Assembly assembly = ExecutionContext.LoadAssembly(name, null, out ignored);
+                Assembly assembly = ExecutionContext.LoadAssembly(name, result, out ignored);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: If we already got the resolved assembly path, we can skip the loading, right? The loading is solely for getting the path of the assembly. When the target assembly is in GAC, we have to load it in order to get the path, but not if we already resolved the assembly file path.

Comment (adityapatwardhan) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: In this case, we actually want to load from module base first. In the `Resolve` method, we do not have access to module base so we need to do this here. If there are alternatives, please suggest.

Comment (PaulHigin) [reply src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: Please add a comment to the code so it is clear to other developers that we really want to load the assembly here, not just return the pathname.  Thanks!

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by adityapatwardhan** [comment]
@daxian-dbw Please have another look.

**Comment by anmenaga** [comment]
@SteveL-MSFT @adityapatwardhan Here is example scenario: servicing a multi-version module.
Let's say there is a widely-used module that over time released several versions.
Then someone discovers a security bug in .NET that requires a client-side code fix from module authors.
It's a lot easier to put a fixed DLL into GAC than try to update DLL in all module version folders.

@adityapatwardhan is there an issue for this PR? did somebody complain?

**Comment by SteveL-MSFT** [comment]
@anmenaga the real world scenario is that we have Windows partners that also ship their modules on PSGallery.  The PSGallery version that the user explicitly installed may be newer than the one that is inbox.  Currently, importing the module loads the older assembly in the GAC which doesn't match the updated module.

**Comment by anmenaga** [comment]
@SteveL-MSFT That's a valid scenario.
Though ideally, I think, the load order should be configurable. Is it something that can be placed in a config file?

**Comment by SteveL-MSFT** [comment]
@anmenaga The updated one in the GAC may not be compatible with modules that were tested against a specific version.  I think this is the tradeoff of app-local versions of assemblies.  At this time, I don't think we should solve this problem universally.

**Comment by adityapatwardhan** [comment]
@TravisEz13 Assigning this to you as I would like to get this in the next release.

**Pull request merged**